Enable IlcTrimMetadata by default

MichalStrehovsky opened the pull request:
Enables more aggressive metadata stripping by default. In this mode, the AOT compiler only generates reflection metadata for artifacts that were deemed reflection-visible by either dataflow analysis, or user inputs. Previously we would consider everything that got generated reflection-visible.

* Change the defaults within the compiler (making the conservative mode opt-in).
* Add tests for scenarios in https://github.com/dotnet/runtimelab/issues/656 (resolves dotnet/runtimelab#656).
* Update tests

MichalStrehovsky committed:
Enable IlcTrimMetadata by default

Enables more aggressive metadata stripping by default. In this mode, the AOT compiler only generates reflection metadata for artifacts that were deemed reflection-visible by either dataflow analysis, or user inputs. Previously we would consider everything that got generated reflection-visible.

* Change the defaults within the compiler (making the conservative mode opt-in).
* Add tests for scenarios in https://github.com/dotnet/runtimelab/issues/656 (resolves dotnet/runtimelab#656).
* Update tests

jkotas commented on src/tests/nativeaot/SmokeTests/Dataflow/Dataflow.cs:
```diff
@@ -577,7 +580,8 @@ public static void Run()
-            object o = Marshal.PtrToStructure(pBytes, typeof(ClassWithLayout2));
+            // https://github.com/dotnet/runtime/issues/70072
+            //object o = Marshal.PtrToStructure(pBytes, typeof(ClassWithLayout2));
             Marshal.DestroyStructure(pBytes, typeof(ClassWithLayout3));
```
This line needs to be commented out too. You do not want to be calling DestroyStructure on uninitialized memory

MichalStrehovsky replied:
`ClassWithLayout3` is blittable, so the destroy structure stub is a no-op.

(Notice we also call `PtrToStructure` with `ClassWithLayout2`, not `3`. It's misuse either way and the test relies on the fact that these stubs don't do much. We need to use different type for each call, otherwise we wouldn't notice if one of the intrinsics regresses. I don't expect this to be commented out for long. I just didn't want to fix it because I knew Vitek has a huge change in #70204 that the fix would conflict with and I didn't want to annoy him.)

davidfowl commented on src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.targets:
```diff
@@ -238,7 +238,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <IlcArg Condition="$(NoWarn) != ''" Include='--nowarn:"$([MSBuild]::Escape($(NoWarn)))"' />
       <IlcArg Condition="$(TrimmerSingleWarn) == 'true'" Include="--singlewarn" />
-      <IlcArg Condition="$(IlcTrimMetadata) == 'true'" Include="--reflectedonly" />
+      <IlcArg Condition="$(IlcTrimMetadata) == 'false'" Include="--reflectableartifacts" />
```
That flag name 😅

MichalStrehovsky replied:
Accepting proposals for a better name. Sometimes I think we should have used hyphens to separate words in command line arguments, but I'm not sure that's the only thing missing here.

Btw, @davidfowl you might be happy to hear that the size of the default `dotnet new webapi` build with all default settings and `<PublishAot>true</PublishAot>` added goes down by 40% after this change.

But it's [broken](https://twitter.com/MStrehovsky/status/1532652675879534593) both before and after and I wasn't in a good enough mental state to debug this yet. With DI the problem and the manifestation of the problem is always _very far_ from each other, which is why I love DI so much.

davidfowl replied:
Yea, I'm not in love with the naming but I think the MSBuild property (if there is one) is more important so I wont bikeshed here.

> Btw, @davidfowl you might be happy to hear that the size of the default dotnet new webapi build with all default settings and <PublishAot>true</PublishAot> added goes down by 40% after this change.

Music to my ears 😄 

> But it's [broken](https://twitter.com/MStrehovsky/status/1532652675879534593) both before and after and I wasn't in a good enough mental state to debug this yet. With DI the problem and the manifestation of the problem is always very far from each other, which is why I love DI so much.

This isn't related to DI, it's because of how application parts work in MVC. I can ping you an explain how it works and why that exception is happenining.

MichalStrehovsky replied:
> This isn't related to DI, it's because of how application parts work in MVC. I can ping you an explain how it works and why that exception is happenining.

I'll probably take you up on that when I sit down to investigate. My current theory is that ASP.NET relies on the fact that by default we set `<TrimmerDefaultAction>copy</TrimmerDefaultAction>` (keep assemblies that are not manifested as trimmable even if nothing references them) and NativeAOT currently has an incompatiblity where it treats `copy` as `copyused` (i.e. we still delete assemblies that are not referenced from anywhere). Can't confirm it because attempting to get a repro without NativeAOT hits an illink bug (#2820).

On the runtime team we would eventually like to see `TrimmerDefaultAction` to not be set to `copy` by default (maybe at least `copyused`, or even `link` by default, so that we can actually trim stuff). If my theory is true, it might be nice to do something on the ASP.NET side so that this pattern isn't broken when we start trimming more.

davidfowl replied:
There's a build time task that looks at all referenced assemblies (transitively) that reference Microsoft.AspNetCore.Mvc and we store an assembly attribute in the project's assembly with the name of the referenced assembly (these can also be added dynamically but I'll ignore that for now, yay plugins). At startup we get these attributes, load the assembly and start looking for "application parts". These parts can contain, controllers, compiled razor views, etc but for WebAPI the 90% case is looking for controller types in these assemblies at runtime.

This is an example of what gets burned into the default project for web api project:

<img width="430" alt="image" src="https://user-images.githubusercontent.com/95136/172080988-e794334b-276b-4f4b-b465-11a756f6183b.png">

This pattern is pretty incompatible with trimming at the moment.

kant2002 replied:
> Accepting proposals for a better name.

maybe `—keepadditionalmetadata`? New name even more confusing then previous one.

MichalStrehovsky replied:
> > Accepting proposals for a better name.
> 
> maybe `—keepadditionalmetadata`? New name even more confusing then previous one.

What this switch does is that it generates reflection metadata for any native artifacts (methods, fields, etc.) that were somehow generated (e.g. there was a call to a method, etc.). I would like the switch name to somehow capture that.

kant2002 replied:
Maybe `keepreflectedartifacts` to indicate that you keep them. Also do you have other word instead of `artifact` that’s what confuse me the most. I cannot get context for that word from reading the flag name. Maybe `fullreflectionartifacts`? Just throwing ideas around.

Suchiman replied:
how about `--reflectiondata=all(generated/compiled)|only(statically)Used|none`

kant2002 replied:
I like `--reflectiondata=all(generated/compiled)|only(statically)Used|none` since that that's much easier to understand what's going on, and should play nice with `more usable reflection-free mode` https://github.com/dotnet/runtime/issues/67193

vitek-karas replied:
@davidfowl So if we teach trimmer that when it sees `ApplicationPartAttribute` it should keep the assembly it points to, it should work? Does it mean it needs to keep the entire assembly, or just parts from it?

Side note: since it already does sort of "source gen" (injects attributes), it might as well do the whole source gen and avoid reflection at runtime. But I don't know enough about this to tell if it's even possible.

MichalStrehovsky replied:
> how about `--reflectiondata=all(generated/compiled)|only(statically)Used|none`

We have a winner! Thanks!

MichalStrehovsky commented:
Marking NO MERGE. This has failures in the CI that are hidden because of #70223.

The failures look to be in the tune of `System.Exception : Couldn't find Utf16Utility type in System.Private.CoreLib.` which is kind of expected given dataflow doesn't model `Assembly` APIs:

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Runtime/tests/System/Text/Unicode/Utf16UtilityTests.ValidateChars.cs#L251

Need to look through all the failures and fix up tests. Hopefully none of the failures are "real".

MichalStrehovsky committed:
Merge branch 'main' into reflectedOnly

MichalStrehovsky committed:
FB

MichalStrehovsky commented:
Hm, System.Runtime tests are now clean, but this still has an issue in System.Collections tests. It's not super related to metadata trimming because I can hit this without it:

```csharp
class Program
{
    public static Penguin pingu;

    static void Main()
    {
        typeof(Program).GetField(nameof(pingu)).GetValue(null);
    }
}

class Penguin { }
```

Crashes with:

```
Unhandled Exception: System.TypeLoadException: The type 'Penguin' cannot be found in assembly 'repro, Version=7.0.0.0, Culture=neutral, PublicKeyToken=null'.
   at System.Reflection.Runtime.General.TypeResolver.Resolve(Handle, MetadataReader, TypeContext) + 0x93
   at System.Reflection.Runtime.FieldInfos.NativeFormat.NativeFormatRuntimeFieldInfo.get_FieldRuntimeType() + 0x93
   at System.Reflection.Runtime.FieldInfos.RuntimeFieldInfo.get_FieldType() + 0x6d
   at System.Reflection.Runtime.FieldInfos.NativeFormat.NativeFormatRuntimeFieldInfo.TryGetFieldAccessor() + 0x9e
   at System.Reflection.Runtime.FieldInfos.RuntimeFieldInfo.get_FieldAccessor() + 0x1c8
   at System.Reflection.Runtime.FieldInfos.RuntimeFieldInfo.GetValue(Object) + 0x2d
   at Program.Main() + 0x61
   at reproNative!<BaseAddress>+0x4ac11f
   at reproNative!<BaseAddress>+0x4ac197
```

This will be another pull request we need to get in before this can be merged.

MichalStrehovsky committed:
Merge branch 'main' into reflectedOnly

MichalStrehovsky committed:
Fix PtrToStructure

MichalStrehovsky committed:
Merge branch 'main' into reflectedOnly

MichalStrehovsky committed:
Merge from main

MichalStrehovsky commented:
Sigh, still cannot be merged.

So now we have a situation where:

* We know `Dictionary<int,int>._entries` needs to be reflection gettable/settable
* We make sure the field gets metadata, we have runtime mapping for the instantiation
* We also make sure to generate a type handle for `Dictionary.Entry<int,int>[]` (the type of the field)
* But as the reflection stack is resolving the signature of the array type, it needs a type handle for the element. We don't have that because it was optimized away. Sigh.

I don't want us to be generating constructed type handles for the element types of arrays as a blanket rule because it's a waste. We have a MethodTable for the element type, but it's not constructed and reflection is not allowed to see it.

I think it is time to allow reflection to see unconstructed MethodTables.

MichalStrehovsky committed:
Consider some array element types constructed

MichalStrehovsky committed:
Fix the tests reflecting on collection privates

MichalStrehovsky commented:
> I don't want us to be generating constructed type handles for the element types of arrays as a blanket rule because it's a waste.

I was wrong. This is observable. Fixed the dependency analysis to account for that and added a test.

MichalStrehovsky committed:
Merge branch 'main' into reflectedOnly

MichalStrehovsky committed:
Keep referenced fields on System.Attribute

MichalStrehovsky committed:
Fix one more test that reflects on implementation details

MichalStrehovsky committed:
Fix the tests

More reflection that only worked by accident. The code is not trim-safe.

MichalStrehovsky commented:
Cool, might be able to finally merge this after all the yak shaving pull requests and "one last commit" commits.

* Hello World with all default settings is now 22% smaller compared to what we just shipped in Preview 5.
* We're at 3.75 MB for a selfcontained EXE Hello World on Windows. It's still not perfect yet, but I'll take it. It can be tuned more with feature switches and other compiler settings. But we'll need to decide which compiler settings we want to support officially, if any.

Cc @dotnet/ilc-contrib
